Bad interaction between never_type, try_blocks, and From/Into

The following code does not compile but the 2 commented alternatives compile:

```rust
#![feature(never_type)]
#![feature(try_blocks)]

fn bar(_: Result<impl Into<!>, u32>) {
// fn bar(_: Result<impl Into<()>, u32>) {
// fn bar(_: Result<!, u32>) {
    unimplemented!()
}

pub fn foo(x: Result<!, u32>) {
    bar(try { x? });
}
```

The error is:

```
error[E0277]: the trait bound `!: From<()>` is not satisfied
  --> src/lib.rs:11:9
   |
11 |     bar(try { x? });
   |     --- ^^^^^^--^^
   |     |   |     |
   |     |   |     this tail expression is of type `Result<(), u32>`
   |     |   the trait `From<()>` is not implemented for `!`, which is required by `(): Into<!>`
   |     required by a bound introduced by this call
   |
   = note: this error might have been caused by changes to Rust's type-inference algorithm (see issue #48950 <https://github.com/rust-lang/rust/issues/48950> for more information)
   = help: did you intend to use the type `()` here instead?
   = note: required for `()` to implement `Into<!>`
```

For some reason, rustc believes that the try block should return `()` instead of `!`.

### Meta

`rustc --version --verbose`:
```
rustc 1.80.0-nightly (7d83a4c13 2024-05-06)
binary: rustc
commit-hash: 7d83a4c131ab9ae81a74c6fd825c827d74a2881d
commit-date: 2024-05-06
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.4
```

--- a/tests/ui/never_type/try-block-never-type-fallback.e2021.stderr
+++ b/tests/ui/never_type/try-block-never-type-fallback.e2021.stderr
@@ -1,0 +1,20 @@
+error[E0277]: the trait bound `!: From<()>` is not satisfied
+  --> $DIR/try-block-never-type-fallback.rs:20:9
+   |
+LL |     bar(try { x? });
+   |     --- ^^^^^^^^^^ the trait `From<()>` is not implemented for `!`
+   |     |
+   |     required by a bound introduced by this call
+   |
+   = note: this error might have been caused by changes to Rust's type-inference algorithm (see issue #148922 <https://github.com/rust-lang/rust/issues/148922> for more information)
+   = help: you might have intended to use the type `()` here instead
+   = note: required for `()` to implement `Into<!>`
+note: required by a bound in `bar`
+  --> $DIR/try-block-never-type-fallback.rs:15:23
+   |
+LL | fn bar(_: Result<impl Into<!>, u32>) {
+   |                       ^^^^^^^ required by this bound in `bar`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/never_type/try-block-never-type-fallback.rs
+++ b/tests/ui/never_type/try-block-never-type-fallback.rs
@@ -1,0 +1,25 @@
+//@ revisions: e2021 e2024
+//@[e2021] edition: 2021
+//@[e2024] edition: 2024
+//@[e2024] check-pass
+
+// Issue #125364: Bad interaction between never_type, try_blocks, and From/Into
+//
+// In edition 2021, the never type in try blocks falls back to (),
+// causing a type error (since (): Into<!> does not hold).
+// In edition 2024, it falls back to !, allowing the code to compile correctly.
+
+#![feature(never_type)]
+#![feature(try_blocks)]
+
+fn bar(_: Result<impl Into<!>, u32>) {
+    unimplemented!()
+}
+
+fn foo(x: Result<!, u32>) {
+    bar(try { x? });
+    //[e2021]~^ ERROR the trait bound `!: From<()>` is not satisfied
+}
+
+fn main() {
+}

